compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala

```diff
@@ -337,37 +337,52 @@ class CheckCaptures extends Recheck, SymTransformer:
     /** If `tpt` is an inferred type, interpolate capture set variables appearing contra-
      *  variantly in it. Also anchor Fresh instances with anchorCaps.
      */
-    private def interpolateVarsIn(tpt: Tree, sym: Symbol)(using Context): Unit =
+    private def interpolateIfInferred(tpt: Tree, sym: Symbol)(using Context): Unit =
       if tpt.isInstanceOf[InferredTypeTree] then
-        interpolator().traverse(tpt.nuType)
-          .showing(i"solved vars in ${tpt.nuType}", capt)
+        interpolate(tpt.nuType, sym)
+          .showing(i"solved vars for $sym in ${tpt.nuType}", capt)
         anchorCaps(sym).traverse(tpt.nuType)
-      for msg <- ccState.approxWarnings do
-        report.warning(msg, tpt.srcPos)
-      ccState.approxWarnings.clear()
+        for msg <- ccState.approxWarnings do
+          report.warning(msg, tpt.srcPos)
+        ccState.approxWarnings.clear()
 
     /** Assert subcapturing `cs1 <: cs2` (available for debugging, otherwise unused) */
     def assertSub(cs1: CaptureSet, cs2: CaptureSet)(using Context) =
       assert(cs1.subCaptures(cs2).isOK, i"$cs1 is not a subset of $cs2")
 
     /** If `res` is not CompareResult.OK, report an error */
-    def checkOK(res: CompareResult, prefix: => String, added: CaptureRef | CaptureSet, pos: SrcPos, provenance: => String = "")(using Context): Unit =
+    def checkOK(res: CompareResult, prefix: => String, added: CaptureRef | CaptureSet, target: CaptureSet, pos: SrcPos, provenance: => String = "")(using Context): Unit =
       res match
         case res: CompareFailure =>
-          inContext(root.printContext(added, res.blocking)):
+          def msg(provisional: Boolean) =
             def toAdd: String = errorNotes(res.errorNotes).toAdd.mkString
             def descr: String =
               val d = res.blocking.description
               if d.isEmpty then provenance else ""
-            report.error(em"$prefix included in the allowed capture set ${res.blocking}$descr$toAdd", pos)
+            def kind = if provisional then "previously estimated\n" else "allowed "
+            em"$prefix included in the ${kind}capture set ${res.blocking}$descr$toAdd"
+          target match
+            case target: CaptureSet.Var
+            if res.blocking.isProvisionallySolved =>
+              report.warning(
+                msg(provisional = true)
+                  .prepend(i"Another capture checking run needs to be scheduled because\n"),
```

Comment:
Do we want to report this warning to end users?

Reply:
For now it's good to have it since we have not tested the recursion very well yet. We will probably remove it later and make it a debug message.

Reply:
OK